lib/pgBackRest/Common/Http/Client.pm

```diff
@@ -0,0 +1,284 @@
+            {name => 'lBufferMax', optional => true, default => 32768, trace => true},
+            {name => 'bVerifySsl', optional => true, default => true, trace => true},
+        );
+
+    # Connect to the server
+    my $oSocket = IO::Socket::SSL->new(
+        PeerHost => $strHost, PeerPort => 'https', SSL_verify_mode => $bVerifySsl ? SSL_VERIFY_PEER : SSL_VERIFY_NONE);
+
+    if (!defined($oSocket))
+    {
+        logErrorResult(ERROR_PROTOCOL, "unable to connect $!", $SSL_ERROR);
+    }
+
+    # Create the buffered IO object
+    my $self = new pgBackRest::Common::Io::Buffered(
+        new pgBackRest::Common::Io::Handle('httpClient', $oSocket, $oSocket), $iProtocolTimeout, $lBufferMax);
+
+    # Bless with the class
+    @ISA = $self->isA();                                            ## no critic (ClassHierarchies::ProhibitExplicitISA)
+    bless $self, $class;
+
+    # Store socket
+    $self->{oSocket} = $oSocket;
+
+    # Generate the query string
+    my $strQuery = httpQuery($hQuery);
+
+    # Construct the request headers
+    $self->{strRequestHeader} = "${strVerb} ${strUri}?${strQuery} HTTP/1.1" . "\r\n";
+
+    foreach my $strHeader (sort(keys(%{$hRequestHeader})))
+    {
+        $self->{strRequestHeader} .= "${strHeader}: $hRequestHeader->{$strHeader}\r\n";
+    }
+
+    $self->{strRequestHeader} .= "\r\n";
+
+    # Write request headers
+    $self->write(\$self->{strRequestHeader});
+
+    # Write content
+    if (defined($rstrRequestBody))
+    {
+        my $iTotalSize = length($$rstrRequestBody);
+        my $iTotalSent = 0;
+
+        # Write the request body in buffer-sized chunks
+        do
+        {
+            my $strBufferWrite = substr($$rstrRequestBody, $iTotalSent, $lBufferMax);
+            $iTotalSent += $self->write(\$strBufferWrite);
+        } while ($iTotalSent < $iTotalSize);
+    }
+
+    # Read response code
+    ($self->{strResponseProtocol}, $self->{iResponseCode}, $self->{strResponseMessage}) = split(' ', trim($self->readLine()));
+
+    # Read the response headers
+    $self->{iContentLength} = undef;
+
+    my $strResponseHeader = '';
+    my $strHeader = trim($self->readLine());
+
+    while ($strHeader ne '')
+    {
+        # Validate header
+        $strResponseHeader .= "${strHeader}\n";
+
+        my $iColonPos = index($strHeader, ':');
+
+        if ($iColonPos == -1)
+        {
+            confess &log(ERROR, "http header '${strHeader}' requires colon separator", ERROR_PROTOCOL);
+        }
+
+        # Parse header
+        my $strHeaderKey = lc(substr($strHeader, 0, $iColonPos));
+        my $strHeaderValue = trim(substr($strHeader, $iColonPos + 1));
+
+        # Store the header
+        $self->{hResponseHeader}{$strHeaderKey} = $strHeaderValue;
+
+        # Process content length
+        if ($strHeaderKey eq HTTP_HEADER_CONTENT_LENGTH)
+        {
+            $self->{iContentLength} = $strHeaderValue + 0;
```

Comment:
Not sure I follow. This header contains the length as an integer of the content that will follow after the header block.  It should be read as an integer.